lib/internal/fs/glob.js

```diff
@@ -0,0 +1,202 @@
+'use strict';
+const { lstatSync, readdirSync } = require('fs');
+const { join, resolve } = require('path');
+
+const {
+  kEmptyObject,
+} = require('internal/util');
+const { isRegExp } = require('internal/util/types');
+const {
+  validateFunction,
+  validateObject,
+} = require('internal/validators');
+
+const {
+  ArrayPrototypeForEach,
+  ArrayPrototypeMap,
+  ArrayPrototypeFlatMap,
+  ArrayPrototypePop,
+  ArrayPrototypePush,
+  SafeMap,
+  SafeSet,
+} = primordials;
+
+let minimatch;
+function lazyMinimatch() {
+  minimatch ??= require('internal/deps/minimatch/index');
+  return minimatch;
+}
+
+function testPattern(pattern, path) {
+  if (pattern === lazyMinimatch().GLOBSTAR) {
+    return true;
+  }
+  if (typeof pattern === 'string') {
+    return true;
+  }
+  if (typeof pattern.test === 'function') {
+    return pattern.test(path);
+  }
+}
+
+class Cache {
+  #caches = new SafeMap();
+  #statsCache = new SafeMap();
+  #readdirCache = new SafeMap();
+
+  stats(path) {
+    if (this.#statsCache.has(path)) {
+      return this.#statsCache.get(path);
+    }
+    let val;
+    try {
+      val = lstatSync(path);
+    } catch {
+      val = null;
+    }
+    this.#statsCache.set(path, val);
+    return val;
+  }
+  readdir(path) {
+    if (this.#readdirCache.has(path)) {
+      return this.#readdirCache.get(path);
+    }
+    let val;
+    try {
+      val = readdirSync(path, { withFileTypes: true });
+      ArrayPrototypeForEach(val, (dirent) => this.#statsCache.set(join(path, dirent.name), dirent));
+    } catch {
+      val = [];
+    }
+    this.#readdirCache.set(path, val);
+    return val;
+  }
+
+  seen(pattern, index, path) {
+    return this.#caches.get(path)?.get(pattern)?.has(index);
+  }
+  add(pattern, index, path) {
+    if (!this.#caches.has(path)) {
+      this.#caches.set(path, new SafeMap([[pattern, new SafeSet([index])]]));
+    } else if (!this.#caches.get(path)?.has(pattern)) {
+      this.#caches.get(path)?.set(pattern, new SafeSet([index]));
+    } else {
+      this.#caches.get(path)?.get(pattern)?.add(index);
+    }
+  }
+
+}
+
+function glob(patterns, options = kEmptyObject) {
+  validateObject(options, 'options');
+  const root = options.cwd ?? '.';
+  const { exclude } = options;
+  if (exclude != null) {
+    validateFunction(exclude, 'options.exclude');
+  }
+
+  const { Minimatch, GLOBSTAR } = lazyMinimatch();
+  const results = new SafeSet();
+  const matchers = ArrayPrototypeMap(patterns, (pattern) => new Minimatch(pattern));
```

Comment:
May want to consider passing some options to the minimatch constructor.  Something like this:

```suggestion
  const matchers = ArrayPrototypeMap(patterns, (pattern) => new Minimatch(pattern, {
    __proto__: null,
    nocase: process.platform === 'win32' || process.platform === 'darwin',
    nonegate: true,
    nocomment: true,
    optimizationLevel: 2,
    platform: process.platform,
  }));
```

Reply:
May also consider `windowsPathsNoEscape` on Windows systems, as it'll likely be confusing on Windows to have `test\*.js` try to run the file named `'test*.js'`, since tab-completion expands to `\` for directories in cmd.exe and powershell.